packages/message-relayer/src/relay-tools.ts

```diff
@@ -0,0 +1,113 @@
+/* Imports: External */
+import { Contract, ethers } from 'ethers'
+import { getContractInterface } from '@eth-optimism/contracts'
+
+/* Imports: Internal */
+import { SentMessage } from './types'
+
+/**
+ * Parses a SentMessage event into a nice SentMessage struct.
+ * @param event SentMessage event to parse.
+ * @param l2BlockOffset The offset between the transaction index and the L2 block number.
+ * @returns Parsed SentMessage event.
+ */
+export const parseSentMessageEvent = (
+  event: ethers.Event,
+  l2BlockOffset: number
+): SentMessage => {
+  // Event signature is `event SentMessage(bytes message)`
+  // See here for reference:
+  // https://github.com/ethereum-optimism/optimism/blob/2d352de8f257db2679b8ac6f953c146a8e99f05d/packages/contracts/contracts/optimistic-ethereum/iOVM/bridge/messaging/iAbs_BaseCrossDomainMessenger.sol#L14
+  const message = event.args.message
+  if (message === undefined) {
+    throw new Error('event is not a SentMessage event')
+  }
+
+  // Message is an encoded call to `relayMessage`. We'll want to decode the components of the
+  // message to simplify the process of inspecting the contents. We could theoretically leave this
+  // decoding step out, but it's nice to have more information where possible.
+  let decoded: {
+    _target: string
+    _sender: string
+    _message: string
+    _messageNonce: ethers.BigNumber
+  }
+
+  try {
+    // Events are emitted on L2 here:
+    // https://github.com/ethereum-optimism/optimism/blob/2d352de8f257db2679b8ac6f953c146a8e99f05d/packages/contracts/contracts/optimistic-ethereum/OVM/bridge/messaging/Abs_BaseCrossDomainMessenger.sol#L88
+    // Encoding is performed here:
+    // https://github.com/ethereum-optimism/optimism/blob/2d352de8f257db2679b8ac6f953c146a8e99f05d/packages/contracts/contracts/optimistic-ethereum/OVM/bridge/messaging/Abs_BaseCrossDomainMessenger.sol#L116-L122
+    decoded = getContractInterface(
+      'OVM_L2CrossDomainMessenger'
+    ).decodeFunctionData('relayMessage', message) as any
+  } catch (err) {
+    throw new Error('unable to parse SentMessage event')
```

Comment:
Can we include the txhash in this error message?

Reply:
Done.